src/api/procedures/addInstruction.ts

```diff
@@ -98,128 +118,198 @@ export async function prepareAddInstruction(
+        addAndAffirmInstructions.push([
+          rawVenueId,
+          rawSettlementType,
+          rawTradeDate,
+          rawValueDate,
+          rawLegs,
+          portfoliosToAffirm[i].map(portfolio =>
+            portfolioIdToMeshPortfolioId(portfolioLikeToPortfolioId(portfolio), context)
+          ),
+        ]);
+      } else {
+        addInstructions.push([rawVenueId, rawSettlementType, rawTradeDate, rawValueDate, rawLegs]);
+      }
+    }
+  });
+
+  if (noLegs.length) {
     throw new PolymeshError({
       code: ErrorCode.ValidationError,
       message: "The legs array can't be empty",
+      data: {
+        instructionNumber: noLegs,
+      },
     });
   }
 
-  let endCondition;
-
-  if (endBlock) {
-    const latestBlock = await context.getLatestBlock();
-
-    if (endBlock.lte(latestBlock)) {
-      throw new PolymeshError({
-        code: ErrorCode.ValidationError,
-        message: 'End block must be a future block',
-      });
-    }
-
-    endCondition = { type: InstructionType.SettleOnBlock, value: endBlock } as const;
-  } else {
-    endCondition = { type: InstructionType.SettleOnAffirmation } as const;
+  if (noFutureBlock.length) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'End block must be a future block',
+      data: {
+        instructionNumber: noFutureBlock,
+      },
+    });
   }
 
-  if (tradeDate && valueDate && tradeDate > valueDate) {
+  if (noFutureDate.length) {
     throw new PolymeshError({
       code: ErrorCode.ValidationError,
       message: 'Value date must be after trade date',
+      data: {
+        instructionNumber: noFutureDate,
+      },
     });
   }
 
-  const rawVenueId = numberToU64(venueId, context);
-  const rawSettlementType = endConditionToSettlementType(endCondition, context);
-  const rawTradeDate = tradeDate ? dateToMoment(tradeDate, context) : null;
-  const rawValueDate = valueDate ? dateToMoment(valueDate, context) : null;
-  const rawLegs: {
-    from: PortfolioId;
-    to: PortfolioId;
-    asset: Ticker;
-    amount: Balance;
-  }[] = [];
-
-  await Promise.all(
-    legs.map(async ({ from, to, amount, token }) => {
-      const fromId = portfolioLikeToPortfolioId(from);
-      const toId = portfolioLikeToPortfolioId(to);
-
-      await Promise.all([
-        assertPortfolioExists(fromId, context),
-        assertPortfolioExists(toId, context),
-      ]);
-
-      const rawFromPortfolio = portfolioIdToMeshPortfolioId(fromId, context);
-      const rawToPortfolio = portfolioIdToMeshPortfolioId(toId, context);
-
-      rawLegs.push({
-        from: rawFromPortfolio,
-        to: rawToPortfolio,
-        asset: stringToTicker(typeof token === 'string' ? token : token.ticker, context),
-        amount: numberToBalance(amount, context),
-      });
-    })
-  );
-
-  let newInstruction;
+  let affirmedInstructions: PostTransactionValue<Instruction[]> | undefined;
+  let addedInstructions: PostTransactionValue<Instruction[]>;
```

Comment:
You could just do `let instructions: PostTransactionValue<Instruction[]>;` and use it to store the result of both `adddBatchTransaction` calls